### PR TITLE
Implement reference kernel for Softmax using CMSIS-NN

### PR DESCRIPTION
**Pull request by giorgio-arenarm**

### CODE REVIEW

**Comment by jianlijianli** [comment]
Thanks for the contribution. The main logic is similar to the 8bit quantized softmax code before we switched to look-up-tables ( https://github.com/tensorflow/tensorflow/commit/96b38db29c3ad3b1c1397f57bca85e3df3a1ac5f#diff-77415d2cb3521cc7859a3349382057ad).

Assigning to Pete who knows the micro code better.

**Comment by gmiodice** [comment]
@giorgio-arenarm it looks that the file: cmsis.inc (..... micro/tools/ext_libs) has not been edit. I suppose we should append $(CMSIS_PATH)/CMSIS/NN/Source/SoftmaxFunctions/arm_softmax_s8.c to THIRD_PARTY_CC_SRCS

**Review by freddan80 — DISMISSED** [review]

**Review by petewarden — DISMISSED** [review]

**Comment by giorgio-arenarm** [comment]
Hello @petewarden are internal tests failing here because of my patch?

**Comment by rthadur** [comment]
@giorgio-arenarm here is the internal error we are getting , can you please check this
`Testing SimpleTestQuantizedSigned4D
Only float32 and uint8_t supported currently, got 9.
Only float32 and uint8_t supported currently, got 9.
kTfLiteOk == registration->invoke(&context, &node) failed at third_party/tensorflow/lite/experimental/micro/kernels/softmax_test.cc:212 (0 vs 1)
expected_output_data.begin()[i] == output_data[i] failed at third_party/tensorflow/lite/experimental/micro/kernels/softmax_test.cc:217 (-117 vs -125)`

**Comment by gbaned** [comment]
@giorgio-arenarm  Could you please check above internal error and resolve the conflicts? Thanks!

**Comment by giorgio-arenarm** [comment]
Hi, I have tried to reproduce those internal errors but unfortunately I haven't been able to do so.
The error you get ("Only float32 and uint8_t supported currently") shouldn't even say what it does since in this same commit I changed the error string inside (both) softmax.cc(s) to "Only float32, uint8_t and int8_t supported currently". I tried to rebase and push, let's see how that goes

**Comment by giorgio-arenarm** [comment]
About the mismatch, I do not get that either. Could I ask what board this test runs on? So to check whether the error might be in the MVE portion of the code or not

**Comment by freddan80** [comment]
@petewarden @advaitjain, gentle ping for comment

**Comment by giorgio-arenarm** [comment]
Rebased. @petewarden @advaitjain, gentle ping for review

**Review by petewarden — DISMISSED** [review]

**Comment by giorgio-arenarm** [comment]
Hi @petewarden is this ready to be merged? Thanks :)

File: tensorflow/lite/micro/kernels/cmsis-nn/softmax.cc [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+                    TfLiteSoftmaxParams* params) {
+  const int batch_size = input->dims->data[0];
+  const int input_size = input->dims->data[1];
+  tflite::reference_ops::Softmax(input->data.f, input_size, batch_size,
+                                 params->beta, output->data.f);
+}
+
+void Softmax1DQuantized(const TfLiteTensor* input, TfLiteTensor* output,
+                        TfLiteSoftmaxParams* params, OpData* data) {
+  // TODO(ahentz): this is arguably a dirty trick. Since the implementation
```
Comment (rthadur) [inline tensorflow/lite/micro/kernels/cmsis-nn/softmax.cc]: please remove TODO

File: tensorflow/lite/micro/kernels/cmsis-nn/softmax.cc [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+                     op_params.input_multiplier,
+                     op_params.input_left_shift,
+                     op_params.diff_min,
+                     GetTensorData<int8_t>(output));
+  }
+}
+
+void Softmax2DQuantized(const TfLiteTensor* input, TfLiteTensor* output,
+                        TfLiteSoftmaxParams* params, OpData* data) {
+  // TODO(ahentz): this is arguably a dirty trick. Since the implementation
```
Comment (rthadur) [inline tensorflow/lite/micro/kernels/cmsis-nn/softmax.cc]: please remove TODO

File: tensorflow/lite/micro/kernels/cmsis-nn/softmax.cc [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+  const TfLiteTensor* input = GetInput(context, node, 0);
+  TfLiteTensor* output = GetOutput(context, node, 0);
+
+  OpData local_data_object;
+  OpData* data = &local_data_object;
+  TF_LITE_ENSURE_STATUS(
+      CalculateSoftmaxOpData(context, input, output, params, data));
+
+  // TODO(ahentz): consider an implementation that works for many (all?)
```
Comment (rthadur) [inline tensorflow/lite/micro/kernels/cmsis-nn/softmax.cc]: please remove TODO

**Review by rthadur — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@giorgio-arenarm Can you please check rthadur's comments and keep us posted? Thanks!

**Comment by giorgio-arenarm** [comment]
Done

**Comment by petewarden** [comment]
Sorry for the lack of comment, I missed that this was still open! We actually had a parallel thread going adding int8 support to softmax internally, so I think internally, so I think this PR is now redundant? Apologies for the duplication of work, please re-open if I'm incorrect:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/kernels/softmax.cc

**Comment by giorgio-arenarm** [comment]
Hi @petewarden this is about enabling CMSIS-NN acceleration for softmax int8. Maybe the title is a bit misleading (sorry about that :) ) but it should still be missing. The committed files here are under the cmsis-nn folder.

**Comment by petewarden** [comment]
Sorry about that, and thanks for the correction! I've reopened this now.

**Review by petewarden — COMMENTED** [review]
Looks good, thanks!

**Comment by giorgio-arenarm** [comment]
@rthadur Is this ready to be merged?

**Review by rthadur — APPROVED** [review]

File: tensorflow/lite/micro/kernels/softmax_test.cc [hunk 4]
```diff
@@ -322,4 +322,360 @@ TF_LITE_MICRO_TEST(SimpleTestQuantizedSigned) {
...
+TF_LITE_MICRO_TEST(SimpleTestQuantizedSigned2D) {
+  using tflite::testing::F2QS;
+
+  const float input_min = -63.5f;
+  const float input_max = 64.0f;
+  const float output_min = 0.0f;
+  const float output_max = (255.0f / 256.0f);
+  const int output_dims_count = 5;
+  int8_t output_data[output_dims_count];
+  tflite::testing::TestSoftmaxQuantizedSigned(  //
```
Comment (rthadur) [inline tensorflow/lite/micro/kernels/softmax_test.cc]: `==1387==ERROR: AddressSanitizer: stack-buffer-overflow on address  at pc  bp  sp 
WRITE of size 1 at  thread T0
    #0  in void tflite::reference_integer_ops::Softmax<signed char>(tflite::SoftmaxParams const&, tflite::RuntimeShape const&, signed char const*, tflite::RuntimeShape const&, signed char*) third_party/tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h:93:36
    #1  in tflite::ops::micro::activations::Softmax2DQuantized(TfLiteTensor const*, TfLiteTensor*, TfLiteSoftmaxParams*, tflite::ops::micro::activations::(anonymous namespace)::OpData*) third_party/tensorflow/lite/micro/kernels/softmax.cc:154:7
    #2  in tflite::ops::micro::activations::SoftmaxEval(TfLiteContext*, TfLiteNode*) third_party/tensorflow/lite/micro/kernels/softmax.cc:233:9
    #3  in tflite::testing::(anonymous namespace)::TestSoftmaxQuantizedSigned(std::initializer_list<int>, std::initializer_list<signed char>, float, float, std::initializer_list<signed char>, std::initializer_list<int>, float, float, signed char*) third_party/tensorflow/lite/micro/kernels/softmax_test.cc:212:3
    #4  in main third_party/tensorflow/lite/micro/kernels/softmax_test.cc:366:3
    #5  in __libc_start_main (/usr/grte/v4/lib64/libc.so.6+0x38bbc)
    #6  in _start /usr/grte/v4/debug-src/src/csu/../sysdeps/x86_64/start.S:108

Address  is located in stack of thread T0 at offset 565 in frame
    #0  in main third_party/tensorflow/lite/micro/kernels/softmax_test.cc:225
`

**Review by rthadur — CHANGES_REQUESTED** [review]

**Comment by giorgio-arenarm** [comment]
Hi @rthadur thank you for validating this patch. Could I ask you how you've obtained this result?
I haven't been able to reproduce it just yet, but from the output log it seems to be unrelated to this change.
The failure seems to be happening in third_party/tensorflow/lite/kernels/internal/reference/integer_ops/softmax.h (the reference code provided in tensorflow) which I did not touch in this patch and moreover this seems to be called by third_party/tensorflow/lite/micro/kernels/softmax.cc which - again - shouldn't have anything to do with this patch (the file created here was under the cmsis-nn subfolder rather than just micro/). Thanks!

**Comment by rthadur** [comment]
@giorgio-arenarm those errors are from internal CI tests , @petewarden can you please assist with the above error ?

**Comment by gbaned** [comment]
@giorgio-arenarm Could you please resolve the conflicts? Thanks!

**Comment by giorgio-arenarm** [comment]
Done. Patch rebased, could you please check whether the error has been resolved?

File: tensorflow/lite/micro/kernels/cmsis-nn/softmax.cc [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+                                    const TfLiteSoftmaxParams* params,
+                                    OpData* data) {
+  if (input->type == kTfLiteUInt8 || input->type == kTfLiteInt8) {
+    if (input->type == kTfLiteUInt8) {
+      TF_LITE_ENSURE_EQ(context, output->params.zero_point, 0);
+    } else {
+      TF_LITE_ENSURE_EQ(context, output->params.zero_point, -128);
+    }
+
+    TF_LITE_ENSURE(context, output->params.scale == 1.f / 256);
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/cmsis-nn/softmax.cc]: This should technically be 1/255 (range of -128 to 127 = 255).  I had a look through some of our models, and it looks like they use 1/256 for scale, so perhaps this is an optimization.  I would prefer:
TF_LITE_ENSURE(context, output->params.scale == 1.f / 256 || output->params->scale == 1.f / 255).

File: tensorflow/lite/micro/kernels/softmax_test.cc [hunk 6]
```diff
@@ -322,4 +322,360 @@ TF_LITE_MICRO_TEST(SimpleTestQuantizedSigned) {
...
+      {1, 5},                                   // Input shape.
+      {
+          F2QS(1.0, input_min, input_max),
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/softmax_test.cc]: We have been trying to move away from using F2Q* in favor of CreateQuantizedTensor (see conv_test.cc).  I realize that the rest of this file uses F2Q and F2Qs, so we can leave it as-is.  Just a note for future changes.

**Review by njeffrie — DISMISSED** [review]

File: tensorflow/lite/micro/kernels/softmax_test.cc (hunk 6 above)
Comment (giorgio-arenarm) [reply tensorflow/lite/micro/kernels/softmax_test.cc]: Acknowledged, thanks!

File: tensorflow/lite/micro/kernels/cmsis-nn/softmax.cc (hunk 5 above)
Comment (giorgio-arenarm) [reply tensorflow/lite/micro/kernels/cmsis-nn/softmax.cc]: Done

**Review by njeffrie — DISMISSED** [review]

**Review by rthadur — DISMISSED** [review]

**Comment by advaitjain** [comment]
This PR is failing with the asan builds. 

@giorgio-arenarm, could you take a look and fix the issue?

I was able to patch the PR and reproduce the error locally by building with bazel with this command:
```
CC=clang BAZEL_COMPILER=llvm bazel run --copt=-DADDRESS_SANITIZER  --copt=-fsanitize=address --linkopt=-fsanitize=address tensorflow/lite/micro/kernels:softmax_test_binary
```

Error log:
```
Testing SimpleTestQuantizedSigned2D
=================================================================
==27528==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7fff88d81c25 at pc 0x00000052b69a bp 0x7fff88d7a6a0 sp 0x7fff88d7a698
WRITE of size 1 at 0x7fff88d81c25 thread T0
    #0 0x52b699 in void tflite::reference_integer_ops::Softmax<signed char>(tflite::SoftmaxParams const&, tflite::RuntimeShape const&, signed char const*, tflite::RuntimeShape const&, signed char*) (/usr/local/google/home/advaitjain/.cache/bazel/_bazel_advaitjain/8c393587092df3f3f8b949c07577ca69/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/lite/micro/kernels/softmax_test_binary+0x52b699)
    #1 0x5277ca in tflite::ops::micro::activations::SoftmaxEval(TfLiteContext*, TfLiteNode*) (/usr/local/google/home/advaitjain/.cache/bazel/_bazel_advaitjain/8c393587092df3f3f8b949c07577ca69/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/lite/micro/kernels/softmax_test_binary+0x5277ca)
    #2 0x4f803b in tflite::testing::(anonymous namespace)::TestSoftmaxQuantizedSigned(std::initializer_list<int>, std::initializer_list<signed char>, float, float, std::initializer_list<signed char>, std::initializer_list<int>, float, float, signed char*) (/usr/local/google/home/advaitjain/.cache/bazel/_bazel_advaitjain/8c393587092df3f3f8b949c07577ca69/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/lite/micro/kernels/softmax_test_binary+0x4f803b)
    #3 0x4f51a1 in main (/usr/local/google/home/advaitjain/.cache/bazel/_bazel_advaitjain/8c393587092df3f3f8b949c07577ca69/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/lite/micro/kernels/softmax_test_binary+0x4f51a1)
    #4 0x7f0873b6bbba in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x26bba)
    #5 0x41f139 in _start (/usr/local/google/home/advaitjain/.cache/bazel/_bazel_advaitjain/8c393587092df3f3f8b949c07577ca69/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/lite/micro/kernels/softmax_test_binary+0x41f139)

Address 0x7fff88d81c25 is located in stack of thread T0 at offset 19301 in frame
    #0 0x4f300f in main (/usr/local/google/home/advaitjain/.cache/bazel/_bazel_advaitjain/8c393587092df3f3f8b949c07577ca69/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/lite/micro/kernels/softmax_test_binary+0x4f300f)

  This frame has 49 object(s):
    [32, 256) 'tensors.i1291'
    [320, 480) 'context.i1292'
    [544, 8752) 'resolver.i1293'
    [9008, 9012) 'builtin_data.i1294'
    [9024, 9032) 'inputs_array_data.i1295'
    [9056, 9064) 'outputs_array_data.i1296'
    [9088, 9160) 'node.i1298'
    [9200, 9536) 'tensors.i'
    [9600, 9760) 'context.i'
    [9824, 18032) 'resolver.i'
    [18288, 18292) 'builtin_data.i'
    [18304, 18312) 'inputs_array_data.i'
    [18336, 18344) 'outputs_array_data.i'
    [18368, 18372) 'temporaries_array_data.i'
    [18384, 18456) 'node.i'
    [18496, 18504) 'error_reporter'
    [18528, 18568) 'output_data'
    [18608, 18620) 'ref.tmp'
    [18640, 18680) 'ref.tmp3'
    [18720, 18732) 'ref.tmp15'
    [18752, 18757) 'output_data35'
    [18784, 18796) 'ref.tmp37'
    [18816, 18821) 'ref.tmp45'
    [18848, 18853) 'ref.tmp66'
    [18880, 18892) 'ref.tmp87'
    [18912, 18922) 'output_data124'
    [18944, 18956) 'ref.tmp126'
    [18976, 18986) 'ref.tmp134'
    [19008, 19018) 'ref.tmp170'
    [19040, 19056) 'agg.tmp205'
    [19072, 19084) 'ref.tmp206'
    [19104, 19109) 'output_data245'
    [19136, 19144) 'ref.tmp247'
    [19168, 19173) 'ref.tmp254'
    [19200, 19205) 'ref.tmp275'
    [19232, 19248) 'agg.tmp295'
    [19264, 19272) 'ref.tmp296'
    [19296, 19301) 'output_data334' <== Memory access at offset 19301 overflows this variable
    [19328, 19340) 'ref.tmp336'
    [19360, 19370) 'ref.tmp344'
    [19392, 19402) 'ref.tmp380'
    [19424, 19440) 'agg.tmp415'
    [19456, 19468) 'ref.tmp416'
    [19488, 19493) 'output_data455'
    [19520, 19540) 'ref.tmp457'
    [19584, 19704) 'ref.tmp463'
    [19744, 19864) 'ref.tmp829'
    [19904, 19920) 'agg.tmp1194'
    [19936, 19956) 'ref.tmp1195'
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow (/usr/local/google/home/advaitjain/.cache/bazel/_bazel_advaitjain/8c393587092df3f3f8b949c07577ca69/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/lite/micro/kernels/softmax_test_binary+0x52b699) in void tflite::reference_integer_ops::Softmax<signed char>(tflite::SoftmaxParams const&, tflite::RuntimeShape const&, signed char const*, tflite::RuntimeShape const&, signed char*)
Shadow bytes around the buggy address:
  0x1000711a8330: f2 f2 f8 f8 f8 f8 f8 f2 f2 f2 f2 f2 f8 f8 f2 f2
  0x1000711a8340: f8 f2 f2 f2 f8 f8 f2 f2 f8 f2 f2 f2 f8 f2 f2 f2
  0x1000711a8350: f8 f8 f2 f2 f8 f8 f2 f2 f8 f8 f2 f2 f8 f8 f2 f2
  0x1000711a8360: f8 f8 f2 f2 00 00 f2 f2 f8 f8 f2 f2 f8 f2 f2 f2
  0x1000711a8370: f8 f2 f2 f2 f8 f2 f2 f2 f8 f2 f2 f2 00 00 f2 f2
=>0x1000711a8380: f8 f2 f2 f2[05]f2 f2 f2 00 04 f2 f2 00 02 f2 f2
  0x1000711a8390: 00 02 f2 f2 00 00 f2 f2 00 04 f2 f2 f8 f2 f2 f2
  0x1000711a83a0: f8 f8 f8 f2 f2 f2 f2 f2 f8 f8 f8 f8 f8 f8 f8 f8
  0x1000711a83b0: f8 f8 f8 f8 f8 f8 f8 f2 f2 f2 f2 f2 f8 f8 f8 f8
  0x1000711a83c0: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f2 f2 f2 f2 f2
  0x1000711a83d0: 00 00 f2 f2 f8 f8 f8 f3 f3 f3 f3 f3 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==27528==ABORTING
```

**Comment by njeffrie** [comment]
Looks like line 365 of softmax_test.cc should be 10.

**Comment by giorgio-arenarm** [comment]
Done. My bad, fixed it!

**Review by advaitjain — APPROVED** [review]

**Comment by advaitjain** [comment]
approved, I'll make sure this gets landed.

**Pull request merged**